### PR TITLE
[WIP] changed confusing message about location permission(#2011)

### LINKED ISSUE #2011: Nearby: confusing message about location permission
**Summary:** 
I use an Android 5.1.1 device. Runtime permissions were introduced only in Android 6. When I use the Nearby feature with Location (not the Location permission for the app, the location (GPS) of the phone) turned off, I get an error stating "GPS is disabled" error which is fine. If I give "Cancel" instead of "Enable GPS", I'm shown a error message stating "Nearby places" can't be displayed without location permission. This seem irrelevant as there's no such things as location permission for the app in my device. It's better to improve the wordings a little in this case.

**Possible solution**
When the user taps on "Cancel" instead of "Enable GPS" for the "Enable GPS" pop-up, showing a full screen error page stating GPS needs to be enabled to use this feature might be helpful. It might also contain a "Refresh" icon or something. Asking to enable GPS permission when the user has denied it doesn't seem to be a nice User experience. If he has did it by mistake the "Refresh" icon should help him.

Just wondering, why don't we also allow users to use Nearby to find places needing pictures by manually specifying a location in Maps? This makes the feature usable even without GPS.

**Steps to reproduce:** 
1. Ensure you're using the app on a device with Android version < 6. Also, ensure GPS is turned off.
2. Open the app
3. Go to Nearby
4. Tap cancel for the "Enable GPS" pop-up

Expected: Either Nearby is usable without enabling GPS (OR) a full-screen error is shown stating requirement of GPS (with Refresh button)
Actual: A weird "Nearby permission" pop-up shows up.

**Device and Android version:** 
Samsung Galaxy J1 Ace, Android 5.1.1 (stock)

**Commons app version:** 
2.8.6

**Screen-shots:** 
Just for reference:
screenshot_2018-06-21-15-30-03
screenshot_2018-06-21-15-30-07

**Would you like to work on the issue?**
No.

### PR DESCRIPTION
**Pull request by sp2710**
Fixes #2011 Nearby: confusing message about location permission 

**What changes did you make and why?**
Added a new Alert Dialog when user taps on "Cancel"

**Tests performed (required)**

Tested on Motorola Moto G5 plus with API level 24

screenshot_1547570243

after tapping on cancel
screenshot_1547570252

If the user clicks on cancel again then the user is redirected to contributions activity

### CODE REVIEW

**Comment by sp2710** [comment]
Sure :)

**Comment by sivaraam** [comment]
It would be nice if this could also be tested with an Android device running Android 5 or less as the issue #2011 was filed mainly because a confusing message about permissions was shown in a device running an Android version which does not support runtime permissions (yeah, I'm not asked to allow or deny when the requests access to my location etc.). I'm trying to test this as I have a device running Android 5.1.1 but it would be great if someone could test this before me :-)

Also, IIUC from the screen shots posted the when the user visits 'Nearby' the dialog in the first screen shot is shown. If he hits on 'Cancel' in it then pop-up in the second one would be shown. Is that right? If yes, I'm not sure how this improves the status quo. Adding another dialog just seems to be worsening it. I guess an improvement would be to avoid drop the first pop-up in the first one. That would reduce the frustration a little.

Also, I had another doubt. Given that #591 has been closed, does asking for the GPS permission as a mandatory requirement make sens? I think it could be made optional which would make Nearby even more useful.

**Comment by sp2710** [comment]
@maskaravivek Done reusing DialogUtil for displaying this dialog.

**Comment by sp2710** [comment]
@sivaraam Actually the screenshots are of an emulator running android 5.1. I had already tested on this emulater and also on my device.
Also I think the dialog should be there to inform the user that the GPS needs to be enabled in order to use that feature.After cancelling again the user is redirected to the contributions fragment and no further dialog appears

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyFragment.java [hunk 1]
```diff
@@ -653,7 +654,27 @@ private void checkGps() {
...
                             })
                     .setNegativeButton(R.string.menu_cancel_upload, (dialog, id) -> {
-                        showLocationPermissionDeniedErrorDialog();
+                        DialogUtil.showAlertDialog(
+                                getActivity(),
+                                null,
+                                getString(R.string.nearby_needs_gps),
+                                getString(R.string.enable_gps),
+                                getString(R.string.cancel),
+                                new Runnable() {
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyFragment.java]: Use lambda

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyFragment.java [hunk 2]
```diff
@@ -653,7 +654,27 @@ private void checkGps() {
...
+                                    @Override
+                                    public void run() {
+                                        Intent callGPSSettingIntent = new Intent(
+                                                android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
+                                        Timber.d("Loaded settings page");
+                                        startActivityForResult(callGPSSettingIntent, 1);
+                                    }
+                                },
+                                new Runnable() {
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyFragment.java]: Use lambda

**Comment by maskaravivek** [comment]
> It would be nice if this could also be tested with an Android device running Android 5 or less

The "GPS dialog" suggests that the app has location permissions(on Android 5 it has it by default) but the app wasn't able to fetch the location. 

If the app has location permissions it checks if the GPS provider is enabled: 

```
locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
```

If the `GPS_PROVIDER` isn't enabled, the GPS dialog is shown. 

So the dialog is not about location permissions. It is trying to check if you have enabled location services on your device or not.

**Comment by sivaraam** [comment]
> Also I think the dialog should be there to inform the user that the GPS needs to be enabled in order to use that feature.

It's might be good to have a dialog that informs the user that GPS is required for the feature. I just don't find how the first dialog which plainly states "GPS is disabled" is useful now. It's better not to have subsequent pop-up dialog. In this particular case, the first dialog seems useless, at least to me. So, it would be nice to drop it.

Just to give an idea. If the first dialog is dropped, the flow should be something like:

1. The user visits 'Nearby'
1. If GPS disabled, he is shown a dialog stating "GPS is needed to use this feature"
1. If he hits 'Cancel' he's taken back to 'Contributions'.
1. If he hits 'Enable GPS', he's taken to 'Nearby'.

Hope that clarifies what I'm trying to say. Let meknow if it doesn't.

**Comment by sp2710** [comment]
@maskaravivek I think only one dialog will be sufficient and we should change the message to something like
>GPS needs to be enabled to use this feature. Would you like to enable it?
Would you like to give your suggestions?

**Comment by domdomegg** [comment]
> @maskaravivek I think only one dialog will be sufficient and we should change the message to something like
> 
> > GPS needs to be enabled to use this feature. Would you like to enable it?
> 
> Would you like to give your suggestions?

I agree with changing the current message. On API 19, this is what happens if GPS is disabled and I open the app:
1. Immediately get message about permissions (without navigating to nearby) (which doesn't make sense as runtime permissions only added in API 23)
2. Get message about GPS being off (which is actually correct (well more like location services but it's close) - possibly wording could be improved to "Location services need to be enabled to use nearby. Would you like to enable them?")
3. Clicking cancel on (2) results in a toast being displayed saying I rejected permissions which is false. I think we don't need a message here at all really as the user will understand that by choosing not to enable location they can't use nearby.

Screenshots:

| 1 | 2 | 3 |
| - | - | - |
| Screenshot_1552835643 | Screenshot_1552835649 | Screenshot_1552835651 |

What I think should happen is:

1. I open the app, and nothing happens yet as haven't tried to view nearby
2. I navigate to nearby by clicking the tab at the top
3. I get the message (2 previously) telling me location services is disabled, and this part of the app needs them enabled
4a. If I enable them, nearby loads correctly
4b. If I keep them disabled, I am returned back to the contributions view

**Comment by sp2710** [comment]
@domdomegg So shall i modify my PR according to it?

**Comment by domdomegg** [comment]
> @domdomegg So shall i modify my PR according to it?

I'd be happy to merge it. The one tricky thing might be only requesting location on viewing nearby (step 1), as I think location is used when it tries to get the nearby notification card to display on top of contributions fragment. If the other bits work I'd be happy to merge that and then step 1 can be done seperately - whatever is easier.

**Comment by sp2710** [comment]
@domdomegg IMO for contributions fragment we can show dialog with content

>Location services need to be enabled to get nearby notification. Would you like to enable them?

And if in nearby we can show this dialog

>Location services need to be enabled to use nearby. Would you like to enable them?

**Comment by domdomegg** [comment]
> Location services need to be enabled to get nearby notification. Would you like to enable them?

I don't know if this will be clear as users might not know what the nearby notification is referring too, especially as this message is likely to be the first thing the user experiences when opening the app. Maybe just say something like 'location services need to be enabled if you want to know about nearby places of interest that lack media on Commons'

**Comment by sp2710** [comment]
Looks nice.
So shall I update my PR according to it ?

**Comment by neslihanturan** [comment]
@sp2710 please do. After a long discussion, this PR seems like ready to be merged after some string refactors.

**Comment by maskaravivek** [comment]
IMO, this is no longer needed after #2925 is merged. We can probably close this PR

**Comment by neslihanturan** [comment]
Sorry I have missed it @maskaravivek

**Pull request closed without merging**